File: libs/domain/src/capabilities.ts
import type {
  RulebenchCapabilityEntryDto,
  RulebenchCapabilityKindDto,
  RulebenchCapabilityManifestDto,
  RulebenchCapabilitySupportDto,
} from "@asha-rulebench/protocol";

export interface RulebenchCapabilitySupportView {
  readonly declared: boolean;
  readonly validationSupported: boolean;
  readonly runtimeExecutable: boolean;
  readonly protocolExposed: boolean;
  readonly liveHostExposed: boolean;
  readonly uiExposed: boolean;
  readonly regressionCovered: boolean;
  readonly durableAcrossRestart: boolean;
  readonly supportLabel: string;
}

export interface RulebenchCapabilityEntryView {
  readonly id: string;
  readonly kind: RulebenchCapabilityKindDto;
  readonly kindLabel: string;
  readonly version: string;
  readonly support: RulebenchCapabilitySupportView;
  readonly evidence: readonly string[];
}

export interface RulebenchRulesetProviderView {
  readonly providerLabel: string;
  readonly rulesetLabel: string;
  readonly compatibilityLabel: string;
  readonly capabilityCount: number;
}

export interface RulebenchCapabilityManifestView {
  readonly manifestId: string;
  readonly manifestVersion: number;
  readonly generatedArtifactSchema: string;
  readonly governedAshaRevision: string;
  readonly operationVocabularyVersion: string;
  readonly effectVocabularyVersion: string;
  readonly protocolLabel: string;
  readonly hostLabel: string;
  readonly recoveryLabel: string;
  readonly providers: readonly RulebenchRulesetProviderView[];
  readonly rulesetLabels: readonly string[];
  readonly packageLabels: readonly string[];
  readonly scenarioCount: number;
  readonly capabilities: readonly RulebenchCapabilityEntryView[];
}

export function projectCapabilityManifest(
  dto: RulebenchCapabilityManifestDto,
): RulebenchCapabilityManifestView {
  return {
    manifestId: dto.manifestId,
    manifestVersion: dto.manifestVersion,
    generatedArtifactSchema: dto.generatedArtifactSchema,
    governedAshaRevision: dto.governedAshaRevision,
    operationVocabularyVersion: dto.operationVocabularyVersion,
    effectVocabularyVersion: dto.effectVocabularyVersion,
    protocolLabel: `${dto.protocolId} v${dto.protocolVersion}`,
    hostLabel: `${dto.host.adapterId} · ${dto.host.storageMode}`,
    recoveryLabel: `Replay: ${dto.host.replayRecoveryMode}; session: ${dto.host.sessionRecoveryMode}`,
    providers: dto.providers.map((provider) => ({
      providerLabel: identityLabel(provider.provider),
      rulesetLabel: identityLabel(provider.ruleset),
      compatibilityLabel: `pipeline ${provider.operationVocabularyVersion} · effects ${provider.effectOperationVocabularyVersion}`,
      capabilityCount: provider.capabilities.length,
    })),
    rulesetLabels: dto.rulesets.map(identityLabel),
    packageLabels: dto.packages.map(identityLabel),
    scenarioCount: dto.scenarios.length,
    capabilities: dto.capabilities.map(projectCapabilityEntry),
  };
}

function projectCapabilityEntry(
  dto: RulebenchCapabilityEntryDto,
): RulebenchCapabilityEntryView {
  return {
    id: dto.id,
    kind: dto.kind,
    kindLabel: capabilityKindLabel(dto.kind),
    version: dto.version,
    support: {
      ...dto.support,
      supportLabel: supportLabel(dto.support),
    },
    evidence: dto.evidence,
  };
}

function identityLabel(identity: Readonly<{ id: string; version: string }>): string {
  return `${identity.id}@${identity.version}`;
}

function capabilityKindLabel(kind: RulebenchCapabilityKindDto): string {
  switch (kind) {
    case "operation":
      return "Operation";
    case "targeting":
      return "Targeting";
    case "policy":
      return "Policy";
    case "content":
      return "Content";
    case "replay":
      return "Replay";
    case "session":
      return "Session";
  }
}

function supportLabel(support: RulebenchCapabilitySupportDto): string {
  if (!support.declared) return "Not declared";
  if (!support.validationSupported) return "Declared only";
  if (!support.runtimeExecutable) return "Validated, not executable";
  if (!support.protocolExposed) return "Rust runtime only";
  if (!support.liveHostExposed) return "Protocol only";
  if (!support.uiExposed) return "Live host only";
  if (!support.regressionCovered) return "UI exposed, regression gap";
  return support.durableAcrossRestart
    ? "Durable and regression covered"
    : "Regression covered, not restart durable";
}

File: rulebench-rs/crates/rulebench-protocol/src/capabilities.rs
use rulebench_rules::{
    CapabilityEntry, CapabilityIdentity, HostCapabilityProfile, RulebenchCapabilityManifest,
    RulesetProviderManifestEntry,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CapabilitySupportDto {
    pub declared: bool,
    pub validation_supported: bool,
    pub runtime_executable: bool,
    pub protocol_exposed: bool,
    pub live_host_exposed: bool,
    pub ui_exposed: bool,
    pub regression_covered: bool,
    pub durable_across_restart: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CapabilityEntryDto {
    pub id: String,
    pub kind: String,
    pub version: String,
    pub support: CapabilitySupportDto,
    pub evidence: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CapabilityIdentityDto {
    pub id: String,
    pub version: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RulesetProviderDto {
    pub provider: CapabilityIdentityDto,
    pub ruleset: CapabilityIdentityDto,
    pub operation_vocabulary_version: String,
    pub effect_operation_vocabulary_version: String,
    pub capabilities: Vec<CapabilityIdentityDto>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HostCapabilityProfileDto {
    pub adapter_id: String,
    pub storage_mode: String,
    pub content_storage_adapter: String,
    pub replay_storage_adapter: String,
    pub replay_recovery_mode: String,
    pub session_recovery_mode: String,
    pub authority_viewer_mode: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RulebenchCapabilityManifestDto {
    pub manifest_id: String,
    pub manifest_version: u32,
    pub generated_artifact_schema: String,
    pub governed_asha_revision: String,
    pub operation_vocabulary_version: String,
    pub effect_vocabulary_version: String,
    pub protocol_id: String,
    pub protocol_version: u32,
    pub host: HostCapabilityProfileDto,
    pub providers: Vec<RulesetProviderDto>,
    pub rulesets: Vec<CapabilityIdentityDto>,
    pub packages: Vec<CapabilityIdentityDto>,
    pub scenarios: Vec<CapabilityIdentityDto>,
    pub capabilities: Vec<CapabilityEntryDto>,
}

impl From<&RulebenchCapabilityManifest> for RulebenchCapabilityManifestDto {
    fn from(value: &RulebenchCapabilityManifest) -> Self {
        Self {
            manifest_id: value.manifest_id.clone(),
            manifest_version: value.manifest_version,
            generated_artifact_schema: value.generated_artifact_schema.clone(),
            governed_asha_revision: value.governed_asha_revision.clone(),
            operation_vocabulary_version: value.operation_vocabulary_version.clone(),
            effect_vocabulary_version: value.effect_vocabulary_version.clone(),
            protocol_id: crate::PROTOCOL_ID.to_string(),
            protocol_version: crate::PROTOCOL_VERSION,
            host: HostCapabilityProfileDto::from(&value.host),
            providers: value
                .providers
                .iter()
                .map(RulesetProviderDto::from)
                .collect(),
            rulesets: value
                .rulesets
                .iter()
                .map(CapabilityIdentityDto::from)
                .collect(),
            packages: value
                .packages
                .iter()
                .map(CapabilityIdentityDto::from)
                .collect(),
            scenarios: value
                .scenarios
                .iter()
                .map(CapabilityIdentityDto::from)
                .collect(),
            capabilities: value
                .capabilities
                .iter()
                .map(CapabilityEntryDto::from)
                .collect(),
        }
    }
}

impl From<&RulesetProviderManifestEntry> for RulesetProviderDto {
    fn from(value: &RulesetProviderManifestEntry) -> Self {
        Self {
            provider: CapabilityIdentityDto::from(&value.provider),
            ruleset: CapabilityIdentityDto::from(&value.ruleset),
            operation_vocabulary_version: value.operation_vocabulary_version.clone(),
            effect_operation_vocabulary_version: value.effect_operation_vocabulary_version.clone(),
            capabilities: value
                .capabilities
                .iter()
                .map(CapabilityIdentityDto::from)
                .collect(),
        }
    }
}

impl From<&CapabilityIdentity> for CapabilityIdentityDto {
    fn from(value: &CapabilityIdentity) -> Self {
        Self {
            id: value.id.clone(),
            version: value.version.clone(),
        }
    }
}

impl From<&CapabilityEntry> for CapabilityEntryDto {
    fn from(value: &CapabilityEntry) -> Self {
        Self {
            id: value.id.clone(),
            kind: value.kind.code().to_string(),
            version: value.version.clone(),
            support: CapabilitySupportDto {
                declared: value.support.declared,
                validation_supported: value.support.validation_supported,
                runtime_executable: value.support.runtime_executable,
                protocol_exposed: value.support.protocol_exposed,
                live_host_exposed: value.support.live_host_exposed,
                ui_exposed: value.support.ui_exposed,
                regression_covered: value.support.regression_covered,
                durable_across_restart: value.support.durable_across_restart,
            },
            evidence: value.evidence.clone(),
        }
    }
}

impl From<&HostCapabilityProfile> for HostCapabilityProfileDto {
    fn from(value: &HostCapabilityProfile) -> Self {
        Self {
            adapter_id: value.adapter_id.clone(),
            storage_mode: value.storage_mode.clone(),
            content_storage_adapter: value.content_storage_adapter.clone(),
            replay_storage_adapter: value.replay_storage_adapter.clone(),
            replay_recovery_mode: value.replay_recovery_mode.clone(),
            session_recovery_mode: value.session_recovery_mode.clone(),
            authority_viewer_mode: value.authority_viewer_mode.clone(),
        }
    }
}

#[cfg(test)]
mod tests {
    use rulebench_rules::{
        assemble_capability_manifest, CapabilityRegistryInput, HostCapabilityProfile,
    };

    use super::*;

    #[test]
    fn mapping_preserves_support_levels_and_protocol_identity() {
        let manifest = assemble_capability_manifest(
            CapabilityRegistryInput {
                providers: Vec::new(),
                rulesets: Vec::new(),
                packages: Vec::new(),
                scenarios: Vec::new(),
                regression_capability_ids: Vec::new(),
            },
            HostCapabilityProfile {
                adapter_id: "test".to_string(),
                storage_mode: "memory".to_string(),
                content_storage_adapter: "none".to_string(),
                replay_storage_adapter: "memory".to_string(),
                replay_recovery_mode: "finalizedArchive".to_string(),
                session_recovery_mode: "none".to_string(),
                authority_viewer_mode: "liveAuthorityReadback".to_string(),
                authored_content_enabled: false,
                exposes_capabilities_through_protocol: true,
                exposes_capabilities_through_live_host: true,
                exposes_capabilities_in_ui: true,
                durable_content: false,
                durable_finalized_replays: false,
                durable_active_sessions: false,
            },
        )
        .expect("manifest is valid");

        let dto = RulebenchCapabilityManifestDto::from(&manifest);
        assert_eq!(dto.protocol_id, crate::PROTOCOL_ID);
        assert_eq!(dto.protocol_version, crate::PROTOCOL_VERSION);
        assert!(dto
            .capabilities
            .iter()
            .any(|capability| capability.id == "operation.damage"));
    }
}
